### DISCUSSION TITLE
BaseSettings source: use runtime arguments in settings sources

### DISCUSSION THREAD

**Discussion by ivan94fi** [pydantic/pydantic#4170] [Question]
Hi,

I am in the process of converting the configuration for one project in my company to Pydantic.

Currently the configuration is based on some JSON files, and I would like to maintain the current JSON files (some minor modifications are allowed) as primary config source.

So I created my BaseSettings subclass and my BaseModel fields, added a JSON source as reported in the documentation, and everything works as expected with a fixed JSON file path:
```python
def json_config_settings_source(settings: BaseSettings) -> Dict[str, Any]:
    """Configure a JSON source."""
    encoding = settings.__config__.env_file_encoding
    return json.loads(Path('config.json').read_text(encoding))

class MyConfig(BaseSettings):
    # various fields ...

    class Config:
        env_file_encoding = "utf-8"

        @classmethod
        def customise_sources(
            cls,
            init_settings,
            env_settings,
            file_secret_settings,
        ):
            return (
                init_settings,
                json_config_settings_source,
                env_settings,
                file_secret_settings,
            )
```

Now I have a problem: how can I customize the `config.json` argument to `json.loads`? Ideally I would like to be able to pass my script a path to a JSON file, which is parsed and validated, then maybe overwrite some of the configurations with environment variables, as needed.

Is this possible with Pydantic?

A similar question was asked [here](https://github.com/samuelcolvin/pydantic/discussions/3027): essentially the proposed solution is to modify the Config.customise_sources class method at runtime, with the needed parameters. Are there any updates?

Thank you

**Comment by samuelcolvin** [comment]
You might be interested in the discussion on https://github.com/pydantic/pydantic-settings/pull/10.

Tldr, we're changing how it works in v2, feedback on the new idea welcome.

**Comment by mikepqr** [comment]
In the meantime, this works (unlike the previous version I posted and deleted, sorry for the churn!) and it has a pretty clean "API" for my use case.

```python
import json
from pathlib import Path
from typing import Any, Dict

from pydantic import BaseSettings
from pydantic.env_settings import SettingsSourceCallable


class Config(BaseSettings):
    x: int
    y: str

    def __init__(self, file=None, *args, **kwargs):
        if file:
            self.insert_file_into_sources(file)
        super().__init__(*args, **kwargs)

    def insert_file_into_sources(self, file):
        def json_config_settings(settings: BaseSettings) -> Dict[str, Any]:
            """Configure a JSON source."""
            encoding = settings.__config__.env_file_encoding
            return json.loads(Path(file).read_text(encoding))

        def customise_sources(
            init_settings: SettingsSourceCallable,
            env_settings: SettingsSourceCallable,
            file_secret_settings: SettingsSourceCallable,
        ) -> tuple[SettingsSourceCallable, ...]:
            return (
                init_settings,
                env_settings,
                json_config_settings,
                file_secret_settings,
            )

        self.Config.customise_sources = customise_sources


a = Config(file="conf_a.json")
b = Config(file="conf_b.json")
nofile = Config(x=99, y="foo")
print(a)
print(b)
print(nofile)
```

```bash
$ tail *.json
==> conf_a.json <==
{
    "x": 1,
    "y": "a"
}

==> conf_b.json <==
{
    "x": 1,
    "y": "b"
}
$ python -m config
x=1 y='a'
x=1 y='b'
x=99 y='foo'
$ y=bar python -m config
$ y=bar python -m config
x=1 y='bar'
x=1 y='bar'
x=99 y='foo'
```

**Comment by markuspi** [comment]
I leveraged the fact that the `settings` parameter in the `json_config_settings` function is actually the partially initialized instance of the to-be constructed class. This allows setting a private attribute `_file` in the constructor that can be accessed by `json_config_settings`:

```python
import json
from typing import Optional, Tuple, Any, Dict
from pathlib import Path

from pydantic import BaseSettings, PrivateAttr
from pydantic.env_settings import SettingsSourceCallable

class MyConfig(BaseSettings):
    x: int
    y: str

    _file: Optional[str] = PrivateAttr()

    def __init__(self, file=None, *args, **kwargs):
        self._file = file
        super().__init__(*args, **kwargs)

    @staticmethod
    def _json_config_settings(settings: "MyConfig") -> Dict[str, Any]:
        if settings._file:
            return json.loads(Path(settings._file).read_text())
        return {}
    
    class Config:
        def customise_sources(
            init_settings: SettingsSourceCallable,
            env_settings: SettingsSourceCallable,
            file_secret_settings: SettingsSourceCallable,
        ) -> Tuple[SettingsSourceCallable, ...]:
            return (
                init_settings,
                env_settings,
                MyConfig._json_config_settings,
                file_secret_settings,
            )

a = MyConfig(file="conf_a.json")
b = MyConfig(file="conf_b.json")
nofile = MyConfig(x=99, y="foo")
print(a)
print(b)
print(nofile)
```

**Reply by maarek** [reply]
Do you know if something like this still works after the changes in  pydantic/pydantic-settings#10 as I do not see a partially materialized model but just the base class?

**Comment by Rocamonde** [comment]
For anyone still interested in finding a solution to this issue, this is the best approach I could find to date:


```python
import pathlib

from pydantic_settings import BaseSettings, InitSettingsSource, PydanticBaseSettingsSource, YamlConfigSettingsSource


class MyBaseSettings(BaseSettings):
    yaml_file: pathlib.Path

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: InitSettingsSource,  # type: ignore[override]
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        return init_settings, env_settings, YamlConfigSettingsSource(cls, init_settings.init_kwargs["yaml_file"])


class MySettings(MyBaseSettings):
    meaning_of_life: int
    ...


settings = MyBaseSettings(yaml_file=pathlib.Path.cwd() / "my_config.yml")
assert settings.meaning_of_life  == 42
```

**Comment by mil-ad** [comment]
Another solution is to slightly abuse `YamlConfigSettingsSource`: instantiate it outside the constructor and pass its content as init values:

```py
class MySettings(BaseSettings):
  foo: int
  bar: float

yaml_config = YamlConfigSettingsSource(MySettings, yaml_file="path/to/yaml/file")
config = MySettings.model_validate(yaml_config())
```
